lib/jormungandr/src/Cardano/Wallet/Jormungandr/Network.hs

```diff
@@ -163,11 +181,60 @@ mkNetworkLayer j = NetworkLayer
 
     }
   where
+    getTipId' = (getTipId j) `mappingError`
+        ErrNetworkTipNetworkUnreachable
+    getBlock' t = (getBlock j t) `mappingError` \case
+        ErrGetBlockNotFound _ ->
+            ErrNetworkTipNotFound
+        ErrGetBlockNetworkUnreachable e ->
+            ErrNetworkTipNetworkUnreachable e
+
     mappingError = flip withExceptT
 
     tailOrEmpty [] = []
     tailOrEmpty (_:xs) = xs
 
+{-------------------------------------------------------------------------------
+                                 State munching
+-------------------------------------------------------------------------------}
+
+-- | A list of blocks and their hashes.
+-- The first block in this sequence is the network tip.
+-- The last block in this sequence is the block of depth _k_,
+-- which is the last unstable block.
+newtype UnstableBlocks = UnstableBlocks
+    { getUnstableBlocks :: Seq (Hash "BlockHeader", Block Tx) }
+    deriving (Show, Eq, Semigroup, Monoid)
+
+-- | Take the tip block header of the unstable blocks, if it exists.
+unstableBlocksTip :: UnstableBlocks -> Maybe BlockHeader
+unstableBlocksTip (UnstableBlocks Empty) = Nothing
+unstableBlocksTip (UnstableBlocks (_ :|> (_, b))) = Just (header b)
+
+-- | Updates the unstable blocks state using the given "fetch" functions.
+updateUnstableBlocks
+    :: Monad m
+    => m (Hash "BlockHeader")
+    -- ^ Fetches tip
+    -> (Hash "BlockHeader" -> m (Block Tx))
+    -- ^ Fetches block
+    -> UnstableBlocks
+    -- ^ Current unstable blocks state
+    -> m UnstableBlocks
+updateUnstableBlocks getTipId' getBlock' ibs = do
+    t <- getTipId'
+    b <- getBlock' t
+    let _bs' = dropAfterSlot b ibs
```

Comment:
This seems slightly incorrect but I suppose it's just temporary to "illustrate" what it could be once we have the intersection. We need to drop all blocks after the intersection, and fetch again missing headers up to the current tip, right? So, that `b` should really by the point of intersection, right ?

Reply:
It should be all fixed now.